### ISSUE TITLE
MAGMA fails to build on Crusher with rocm enabled

### ISSUE THREAD

**Issue by balos1** [xsdk-project/xsdk-issues#202]
Spack files here: https://github.com/xsdk-project/installxSDK/tree/r-0.8.0/platformFiles/crusher/PrgEnv-gnu and https://github.com/xsdk-project/installxSDK/tree/r-0.8.0/platformFiles/crusher/PrgEnv-cray

```
==> magma: Executing phase: 'cmake'
==> [2022-11-09-20:29:29.437439] Module name: gcc/11.2.0
==> [2022-11-09-20:29:29.437606] Package directory variable prefix: GCC
==> [2022-11-09-20:29:29.444097] '/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/cmake-3.21.3-ldjnovu5ttqqaxltrk352yratvegkqae/bin/cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/opt/spack/linux-sles15-zen3/gcc-11.2.0/magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/opt/spack/linux-sles15-zen3/gcc-11.2.0/magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/lib;/autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/opt/spack/linux-sles15-zen3/gcc-11.2.0/magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/lib64;/opt/rocm-5.2.0/hip/lib;/opt/rocm-5.2.0/lib;/opt/rocm-5.2.0/llvm/lib;/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/openblas-0.3.17-u5cqlpli7ttlktg4cx4g45fik332kxzv/lib;/opt/rocm-5.2.0/lib64;/opt/cray/pe/gcc/11.2.0/snos' '-DCMAKE_PREFIX_PATH:STRING=/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/openblas-0.3.17-u5cqlpli7ttlktg4cx4g45fik332kxzv;/opt/rocm-5.2.0/llvm;/opt/rocm-5.2.0/;/opt/rocm-5.2.0/;/opt/rocm-5.2.0/;/opt/rocm-5.2.0/hip;/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/cmake-3.21.3-ldjnovu5ttqqaxltrk352yratvegkqae' '-DCMAKE_INSTALL_PREFIX=/autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/opt/spack/linux-sles15-zen3/gcc-11.2.0/magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js' '-DCMAKE_INSTALL_NAME_DIR:PATH=/autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/opt/spack/linux-sles15-zen3/gcc-11.2.0/magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/lib' '-DBLAS_LIBRARIES=/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/openblas-0.3.17-u5cqlpli7ttlktg4cx4g45fik332kxzv/lib/libopenblas.so' '-DLAPACK_LIBRARIES=/sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/openblas-0.3.17-u5cqlpli7ttlktg4cx4g45fik332kxzv/lib/libopenblas.so' '-DBUILD_SHARED_LIBS=ON' '-DUSE_FORTRAN=yes' '-DMAGMA_ENABLE_HIP=ON' '-DCMAKE_CXX_COMPILER=hipcc' '/tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-src'
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is unknown
-- The Fortran compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /autofs/nccs-svm1_proj/csc326/balos1/crusher/spack-xsdk/lib/spack/env/gcc/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /opt/rocm-5.2.0/bin/hipcc
-- Check for working CXX compiler: /opt/rocm-5.2.0/bin/hipcc - broken
CMake Error at /autofs/nccs-svm1_sw/crusher/spack-envs/base/opt/cray-sles15-zen3/gcc-11.2.0/cmake-3.21.3-ldjnovu5ttqqaxltrk352yratvegkqae/share/cmake-3.21/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/opt/rocm-5.2.0/bin/hipcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_ab91b/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_ab91b.dir/build.make CMakeFiles/cmTC_ab91b.dir/build
    gmake[1]: Entering directory '/tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_ab91b.dir/testCXXCompiler.cxx.o
    /opt/rocm-5.2.0/bin/hipcc    -o CMakeFiles/cmTC_ab91b.dir/testCXXCompiler.cxx.o -c /tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    Can't exec "/opt/rocm-5.2.0/hip/llvm/bin/clang": No such file or directory at /opt/rocm-5.2.0/bin//hipcc.pl line 156.
    Use of uninitialized value $HIP_CLANG_VERSION in pattern match (m//) at /opt/rocm-5.2.0/bin//hipcc.pl line 157.
    Can't exec "/opt/rocm-5.2.0/hip/llvm/bin/clang": No such file or directory at /opt/rocm-5.2.0/bin//hipcc.pl line 161.
    Use of uninitialized value $HIP_CLANG_TARGET in scalar chomp at /opt/rocm-5.2.0/bin//hipcc.pl line 162.
    Use of uninitialized value $HIP_CLANG_VERSION in concatenation (.) or string at /opt/rocm-5.2.0/bin//hipcc.pl line 165.
    Use of uninitialized value $HIP_CLANG_INCLUDE_PATH in concatenation (.) or string at /opt/rocm-5.2.0/bin//hipcc.pl line 191.
    Use of uninitialized value $HIP_CLANG_INCLUDE_PATH in concatenation (.) or string at /opt/rocm-5.2.0/bin//hipcc.pl line 192.
    sh: /opt/rocm-5.2.0/hip/llvm/bin/clang: No such file or directory
    gmake[1]: *** [CMakeFiles/cmTC_ab91b.dir/build.make:78: CMakeFiles/cmTC_ab91b.dir/testCXXCompiler.cxx.o] Error 127
    gmake[1]: Leaving directory '/tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_ab91b/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:9 (project)


-- Configuring incomplete, errors occurred!
See also "/tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeOutput.log".
See also "/tmp/balos1/spack-stage/spack-stage-magma-2.6.2-sbeft3mvhltw3zqec3fx677xiacg74js/spack-build-sbeft3m/CMakeFiles/CMakeError.log".
```

spack-build-out.txt

**Comment by balos1** [comment]
Same error with PrgEnv-cray and cce compiler.

**Comment by luszczek** [comment]
MAGMA release 2.7 is for xSDK 0.8 but this build used version 2.6.2 of MAGMA

**Comment by balay** [comment]
I see  2.7 tarball got released today. Will update xsdk branch to use it.

Edit: pushed the update

**Comment by G-Ragghianti** [comment]
It appears that your "hipcc" (and hipcc.pl) needs to call "clang", but it is not in the expected place.  

```
Can't exec "/opt/rocm-5.2.0/hip/llvm/bin/clang": No such file or directory at /opt/rocm-5.2.0/bin//hipcc.pl line 156.
```

**Comment by tcojean** [comment]
We recently saw the same problem that @G-Ragghianti reports with Ginkgo during manual tests. Somehow `spack load` tends to set incorrect environment settings for the important `ROCM_PATH` variable which causes this sort of error within the very brittle Perl scripts. It sometimes sets it to:
```
ROCM_PATH=/opt/rocm-5.2.0/hip
```

Which should instead be
```
ROCM_PATH=/opt/rocm-5.2.0
```

Can you check if this is the case here as well?

And indeed, in `hipcc.pl` https://github.com/ROCm-Developer-Tools/HIP/blob/rocm-5.2.x/bin/hipcc.pl#L152 it tries to use a variable expanded from `ROCM_PATH` (in https://github.com/ROCm-Developer-Tools/HIP/blob/rocm-5.2.x/bin/hipvars.pm). The correct setting should be `HIPCC=/opt/rocm-5.2.0/llvm/bin/clang`.

**Comment by balay** [comment]
yes - rocm-5.2 changed paths.

https://github.com/spack/spack/issues/32826

There was a bug here - that got fixed

https://github.com/spack/spack/pull/33257

So the following change in spack spec to reflect rocm-5.2 path:
```diff
diff --git a/platformFiles/crusher/PrgEnv-gnu/spack.yaml b/platformFiles/crusher/PrgEnv-gnu/spack.yaml
index be6ad89..6ab181f 100644
--- a/platformFiles/crusher/PrgEnv-gnu/spack.yaml
+++ b/platformFiles/crusher/PrgEnv-gnu/spack.yaml
@@ -235,7 +235,7 @@ spack:
       buildable: false
       externals:
       - spec: hip@5.2.0
-        prefix: /opt/rocm-5.2.0/hip
+        prefix: /opt/rocm-5.2.0
         modules: [rocm/5.2.0]
         extra_attributes:
           compilers:
```

**Comment by v-dobrev** [comment]
I think this issue is now resolved by the updated MAGMA v2.7.0 release tarball, see https://github.com/spack/spack/pull/33814. I tested it on Tioga with ROCm/HIP enabled.
